h/activity/query.py

```diff
@@ -0,0 +1,110 @@
+# -*- coding: utf-8 -*-
+
+"""
+The query parser which converts our subset of the Apache Lucene syntax and
+transforms it into a MultiDict structure that h.api.search understands.
+"""
+
+from __future__ import unicode_literals
+
+from collections import namedtuple
+
+import pyparsing as pp
+from webob.multidict import MultiDict
+
+whitespace = set([
+    "\u0009",  # character tabulation
+    "\u000a",  # line feed
+    "\u000b",  # line tabulation
+    "\u000c",  # form feed
+    "\u000d",  # carriage return
+    "\u0020",  # space
+    "\u0085",  # next line
+    "\u00a0",  # no-break space
+    "\u1680",  # ogham space mark
+    "\u2000",  # en quad
+    "\u2001",  # em quad
+    "\u2002",  # en space
+    "\u2003",  # em space
+    "\u2004",  # three-per-em space
+    "\u2005",  # four-per-em space
+    "\u2006",  # six-per-em space
+    "\u2007",  # figure space
+    "\u2008",  # punctuation space
+    "\u2009",  # thin space
+    "\u200a",  # hair space
+    "\u2028",  # line separator
+    "\u2029",  # paragraph separator
+    "\u202f",  # narrow no-break space
+    "\u205f",  # medium mathematical space
+    "\u3000",  # ideographic space
+])
+
+parser = None
+Match = namedtuple('Match', ['key', 'value'])
+
+
+def parse(q):
+    """Parse a free text, Lucene-like, query string into a MultiDict.
+
+    "user:luke tag:foobar tag:news hello world" is parsed into:
+    ``
+    {
+        "user": "luke",
+        "tag": "foobar",
+        "tag": "news",
+        "any": "hello",
+        "any": "world"
+    }
+    ``
+
+    Supported keys for fields are ``user``, ``group``, ``tag``, ``uri``.
+    Any other search terms will get the key ``any``.
+    """
+    parser = _get_parser()
+    parse_results = parser.parseString(q)
+
+    # The parser returns all matched strings, even the field names, we use a
+    # parse action to turn matches into a key/value pair (Match), but we need
+    # to filter out any other matches that the parser returns.
+    return MultiDict([m for m in parse_results if isinstance(m, Match)])
+
+
+def _get_parser():
+    global parser
+    if parser is None:
+        parser = _make_parser()
+    return parser
+
+
+def _make_parser():
+    # Enable memoizing of the parsing logic
+    pp.ParserElement.enablePackrat()
+
+    word = pp.CharsNotIn(' ')
+    word.skipWhitespace = True
+
+    value = pp.MatchFirst([
+        pp.quotedString.copy().setParseAction(pp.removeQuotes),
+        word
+    ])
```

Comment:
Couple of observations about the preceding five lines:

1. You don't actually seem to use the `whitespace` set at the moment as an argument to `CharsNotIn`, which probably isn't intentional?
2. I can't find any documentation of `skipWhitespace` attr for `CharsNotIn`, so I'd be inclined to use something like `pp.Empty() + pp.CharsNotIn(''.join(whitespace))` instead.
3. This is purely stylistic, but I think it might be clearer to add two `QuotedString` matchers rather than duplicating the `pp.quotedString` singleton. This also removes the need for `pp.removeQuotes`.

All in all, that would result in something like this:

```python
value = pp.MatchFirst([
    pp.QuotedString('"', escchar='\\'),
    pp.QuotedString("'", escchar='\\'),
    pp.Empty() + pp.CharsNotIn(''.join(whitespace)),
])
```

which seems a bit clearer to me?